pyesbulk/__init__.py

```diff
@@ -43,65 +68,115 @@ def _calc_backoff_sleep(backoff):
     return _r.uniform(0, min(b, _MAX_SLEEP_TIME))
 
 
+def _sleep_w_backoff(backoff):
+    time.sleep(_calc_backoff_sleep(backoff))
+
+
 def quiet_loggers():
     """
     A convenience function to quiet the urllib3 and elasticsearch1 loggers.
     """
     logging.getLogger("urllib3").setLevel(logging.FATAL)
-    logging.getLogger(es_logger).setLevel(logging.FATAL)
+    logging.getLogger(_es_logger).setLevel(logging.FATAL)
 
 
-def put_template(es, name, body):
+def put_template(es, name=None, mapping_name=None, body=None):
     """
-    put_template(es, name, body)
+    put_template(es, name, mapping_name, body)
+
+    Updates a given template when the version of the template as
+    stored in the mapping is different from the existing one.
 
     Arguments:
 
         es - An Elasticsearch client object already constructed
         name - The name of the template to use
+        mapping_name - The name of the mapping used in the template
         body - The payload body of the template
 
     Returns:
 
         A tuple with the start and end times of the PUT operation, along
-        with the number of times the operation was retried.
+        with the number of times the operation was retried, and string
+        of keywords indicating note-worthy behavior (currently only
+        "original-no-version", which is returned when the mapping being
+        updated has no version number).
 
         Failure modes are raised as exceptions.
     """
+    assert name is not None and mapping_name is not None and body is not None
     retry = True
     retry_count = 0
+    original_no_version = False
     backoff = 1
     beg, end = time.time(), None
+    try:
+        body_ver = int(body["mappings"][mapping_name]["_meta"]["version"])
+    except KeyError:
+        raise Exception(
+            f"Bad template, {name}: mapping name, '{mapping_name}',"
+            " missing from template"
+        )
     while retry:
+        original_no_version = False  # Initialized twice for proper scope
+        try:
+            tmpl = es.indices.get_template(name=name)
+        except es_excs.ConnectionError:
+            # We retry all connection errors
+            _sleep_w_backoff(backoff)
+            backoff += 1
+            retry_count += 1
```

Comment:
Does it make sense to add a condition on the retry count as well along with a boolean retry?

Reply:
Can you explain more of what you mean?  What behavior would you be changing / fixing by doing that?

Reply:
@npalaska?